Add callback parameters for Stable Diffusion pipelines

jamestiotio opened the pull request:
This PR adds the parameters `callback` and `callback_frequency` to the `__call__` methods of the Stable Diffusion pipelines.

This PR closes #459.

As discussed in the accompanying issue, these callbacks can be helpful for consumers of the `diffusers` API who would like to inspect the current progress of the pipeline.

Instead of defining both `callback` and `img_callback` such as those implemented in the [original Stable Diffusion Repository](https://github.com/CompVis/stable-diffusion/tree/main/ldm/models/diffusion), this implementation follows the more general version of the callback, with the callback function having the following signature:

```python
def callback(
    step: int,
    timestep: np.ndarray,
    latents: torch.FloatTensor,
    image: Union[List[PIL.Image.Image], np.ndarray]
):
    # ...
```

This way, downstream consumers can select whichever parameters they want to use to implement incremental diffusion. This also avoids unnecessary, duplicated parameters.

Consumers can also specify the `callback_frequency` parameter to indicate how often they would like the callback function to be called. If `callback` is defined but `callback_frequency` is not specified, the callback function will be called at every step.

Signed-off-by: James R T <jamestiotio@gmail.com>

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_521). All of your documentation changes will be reflected on that endpoint.

keturn commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -103,6 +103,8 @@ def __call__(
         latents: Optional[torch.FloatTensor] = None,
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
+        callback: Optional[Callable] = None,
```
Specify the signature of the callback:
```suggestion
        callback: Optional[Callable[[int, List[PIL.Image.Image]], None]] = None,
```

(and same goes for where this appears in the other pipelines)

keturn commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -259,6 +273,18 @@ def __call__(
+            # call the callback, if provided
+            if callback is not None:
+                if (callback_frequency is None) or (callback_frequency is not None and i % callback_frequency == 0):
+                    # scale and decode the image latents with vae
+                    current_latents = 1 / 0.18215 * latents
+                    image = self.vae.decode(current_latents).sample
+
+                    image = (image / 2 + 0.5).clamp(0, 1)
+                    image = image.cpu().permute(0, 2, 3, 1).numpy()
```
I think this would be a great thing to extract to a method, as the exact same operations are done just outside the loop for the final results. (`decode_latents` was what I ended up calling it when I did this for similar reasons.)

I know that diffusers has a one-class-per-pipeline philosophy, and here's where I'm hoping that doesn't mean it's also a one-_method_-per-pipeline rule.

keturn commented:
When I wanted to see how things were working under the hood, I did this exact thing. Except since I was snoopy I sent the latents themselves and the timestep, not just the decoded image and step number.

I'm on the fence whether it would be a good idea to make the callback signature more complicated to support more use cases, or whether it's best to do the simpler thing that is what people are interested in 95% of the time.

The other thing to note is that this dovetails with the discussion in #374. We could expose intermediate results by yielding them from a generator, or by sending them back through a callback, but putting both interfaces in the same pipeline would be a mess.

keturn commented:
> I'm on the fence whether it would be a good idea to make the callback signature more complicated to support more use cases, or whether it's best to do the simpler thing that is what people are interested in 95% of the time.

If it's worth including an option to result the latents in the output (as proposed by #506), then it's worth including them in the intermediate results.

deepdiffuser commented:
thanks for making this PR @jamestiotio! this will be useful for us in https://github.com/franksalim/franksalim-imagetools

deepdiffuser commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -103,6 +103,8 @@ def __call__(
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
+        callback: Optional[Callable] = None,
+        callback_frequency: Optional[int] = None,
```
Should this default to `2` or `1` instead of None?

jamestiotio commented:
@keturn @deepdiffuser I have implemented your suggestions and fixes.

> If it's worth including an option to result the latents in the output (as proposed by #506), then it's worth including them in the intermediate results.

Alright, I have refactored the implementation to allow the callback function to take in the timesteps and latents as parameters as well.

> The other thing to note is that this dovetails with the discussion in #374. We could expose intermediate results by yielding them from a generator, or by sending them back through a callback, but putting both interfaces in the same pipeline would be a mess.

Hmm, I am not sure about whether a callback function or a generator would be preferable for this. Personally, I don't mind with either one, but I will leave the decision of which one to implement, and if it is the generator version, how to go about implementing that generator (should we simply yield a tuple containing all 4 parameters?), to the maintainers.

Also, the linked issue contains [a comment](https://github.com/huggingface/diffusers/issues/374#issuecomment-1246749044) that raised the problem of returning intermediate results due to the need of running the safety checker, and because of that, I have refactored the implementation to also run the safety checker on every intermediate image before returning them. While this will slow down the overall execution time whenever a callback function is defined, I believe that this should be an acceptable compromise.

patrickvonplaten commented:
Overall, a bit worried that this is not "bare-bone" / "core-functionality" enough to deserve to be in the main pipelines. If we add more and more of such features the pipelines will quickly explode in terms of complexity. 

Would a community pipeline be fine for you maybe? (working on making community pipelines easier to use with pip diffusers)

Also @patil-suraj @anton-l @pcuenca here

keturn commented:
On "core functionality":

I think it's a feature common in other implementations, it's useful for any kind of interactive application, and "how do I save the intermediate images" is one of the FAQs we get over on Stable Diffusion discord.

Granted, that last bit is _partly_ because some people start with incorrect assumptions how the inference process works and they expect the 20th step of a 100-step task will give them the same thing they see as the 20th step from a 20 step task, but that's not an argument _against_ it. If there are low-cost ways to better enable explainability tools like diffusers-interpret, that sounds like great core functionality.

>  I am not sure about whether a callback function or a generator would be preferable for this.

After building out a [little demo](https://github.com/keturn/sd-progress-demo), I don't know whether I'd argue a generator is a _better_ external interface for it, but it is certainly easier to wrap a generator implementation and make it invoke callbacks than it would be to try to do it the other way around.

jamestiotio commented:
Regarding the concern on "core functionality", I agree with @keturn.

> After building out a [little demo](https://github.com/keturn/sd-progress-demo), I don't know whether I'd argue a generator is a _better_ external interface for it, but it is certainly easier to wrap a generator implementation and make it invoke callbacks than it would be to try to do it the other way around.

Thank you for the demo @keturn, it definitely helped me visualise how a generator version of the pipeline can be used.

Another concern that came to my mind after checking out your demo was backward compatibility. Implementing a generator version **by default** (i.e., without the user specifying some kind of flag to `__call__`) might break current applications, since any calls to `StableDiffusionPipeline(...)` and its derivatives would need to be modified in two major ways:

1. Since a generator object is not subscriptable, users who are not interested in the intermediate outputs would not be able to use this code snippet anymore:

    ```python
    image = pipe(prompt).images[0]
    ```

    Instead, users would need to refactor their implementation by using `yield from` and needlessly iterating through all of the intermediate outputs (and throwing them away). This brings us to the second point.

2. The calls to the pipelines have to be put into a function to utilize the `yield from` keyword. While this might seem trivial since using a function to wrap around the pipeline is highly recommended for all of its usual benefits (modularity, maintainability, etc.), and while production-grade applications will most likely use functions anyway, there might be applications that do not use functions for this. The demo examples in various READMEs of `diffusers` [here](https://github.com/huggingface/diffusers/blob/main/README.md), [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/README.md), and [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/README.md) also do not explicitly put the pipelines in functions (understandably so, since they are simply showing the most basic usage of the pipelines).

Providing an explicit, user-specified flag to `__call__` to indicate that the user is aware of this might solve this issue, but I am not sure if this is desirable. Such a flag would basically make the generator version opt-in. Any thoughts about this?

pcuenca commented:
> Overall, a bit worried that this is not "bare-bone" / "core-functionality" enough

I'm torn about this one. On one hand, it is true that adding too many pieces will imply conditional code, additional arguments, and other forms of complexity. From that point of view, I'd rather have it as a community pipeline. On the other hand, this particular feature might be general enough to support many different use cases, as shown by the community. In particular, explainability and interpretability could benefit from it, like [diffusers-interpret](https://github.com/JoaoLages/diffusers-interpret) shows, or this [puzzling latents visualization](https://twitter.com/jeremyphoward/status/1573063719122501632?s=21&t=VykEG14Zb9GAipGlnkpCWA). A community pipeline would also work fine, but it has a couple of drawbacks too: discoverability, and keeping implementations in sync with the main code.

Other considerations:
- I don't see an easy way to make this compatible with the upcoming Flax implementations.
- What to do with the safety checker.

I think `diffusers` has to be welcoming to tinkerers and at the same time not overwhelming. Is this general enough to deserve to be part of the core? After reflecting a bit, I'm leaning towards _yes_, but I'd love to hear more arguments.

On the design itself, after reading the different versions of the code I have a couple of initial comments:
- It should just be the latents alone, in my opinion. Not all uses require decoding.
- I suggest `callback_frequency` should be called `callback_steps` or something along those lines.
- The generator approach is really interesting, I loved the code. But I think it's hard to keep backwards compatibility and support the most common case (no intermediates are required) without substantially breaking the linearity of the code, which is important for clarity.

I'm happy to do a more thorough review after some more discussion :)

jamestiotio commented:
> - It should just be the latents alone, in my opinion. Not all uses require decoding.
> - I suggest `callback_frequency` should be called `callback_steps` or something along those lines.

@pcuenca I have implemented some of your suggestions above.

> - What to do with the safety checker.

Some of my thoughts on this aspect:

1. If we do not run the safety checker, the safety checker is effectively rendered useless since users can bypass it by extracting the latents from the final step and decoding it themselves to get the final output image.

2. If we run the safety checker, then there are two remaining concerns:
   A. This will require decoding the latents using the VAE and passing it to the safety checker. Since the latents are already decoded anyway, we might as well pass it to the callback function to avoid duplicate effort by users who might want to decode the latents.
   B. What value should we return as the latents if the safety checker detects an image with potential NSFW content? If we simply pass the latents back, then users would be able to bypass the safety checker as well (same scenario as the one described in 1).

ahrm commented:
My two cents on this issue is that a callback functionality is essential for most interactive applications. I had to implement it myself for my [desktop frontend](https://github.com/ahrm/UnstableFusion) and I was very surprised to find out that it is not a built-in feature.

I understand your desire to keep the interface simple, but I would say this functionality is too important not to include. The main problem is that you can not simply subclass `StableDiffusionPipeline` and add this functionality, currently the only way to do it is to copy and modify the source code, which may cause a lot of outdated and fragmented implementations of this functionality in the wild.

One possible solution for this issue without complicating the interface is to have a method, say, `on_iteration_end` which is a `noop` by default but it is called after every iteration. If we have this, the user can simply subclass `StableDiffusionPipeline` and override this method, without complicating the call signature.

patrickvonplaten reviewed (approved):
The PR is good to go for me then! Thanks a lot for your thoughtful comments here. 
Agree that callbacks are general enough to be implemented in the core-pipelines

pcuenca commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -106,6 +107,43 @@ def disable_attention_slicing(self):
         self.enable_attention_slicing(None)
 
+    @torch.no_grad()
+    def decode_latents(self, latents: torch.FloatTensor) -> np.ndarray:
```
Since we are just sending the latents to the callback I think we could return the code inside this function back where it was inside `__call__`.

keturn replied:
but then the callback wouldn't be able to reference this method, and would have to copy-and-paste it if it wants that very common functionality.

(I'm okay with saying that's out of scope for this PR and will be addressed separately as part of https://github.com/huggingface/diffusers/issues/551#issuecomment-1259415393 )

jamestiotio replied:
Hi @pcuenca, @keturn makes a good point here. Should we return `decode_latents` and `run_safety_checker` back inside `__call__` then?

pcuenca replied:
I think we should for now, there's another discussion in another issue about doing some code extraction of some sort, so I think we should just focus on the bare callback functionality here.

jamestiotio replied:
Okay, got it!

pcuenca commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -106,6 +107,43 @@ def disable_attention_slicing(self):
+        return image
+
+    @torch.no_grad()
+    def run_safety_checker(self, image: np.ndarray) -> Tuple[np.ndarray, List[bool]]:
```
Same here, I think we should remove this function and return the code to where it was.

pcuenca commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -119,6 +157,8 @@ def __call__(
         latents: Optional[torch.FloatTensor] = None,
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
+        callback: Optional[Callable[[int, np.ndarray, torch.FloatTensor], None]] = None,
```
```suggestion
        callback: Optional[Callable[[int, int, torch.FloatTensor], None]] = None,
```
I _think_ that `timestep` should always be an int. There's an upcoming PR https://github.com/huggingface/diffusers/pull/637#issue-1385265940 that enforces it, but I believe it should already be the case in `main` after #534 was merged, see https://github.com/huggingface/diffusers/commit/bd8df2da89d99f630e5aa2ddb8f8cb45456561f1#diff-b3769bd64cdb5e99671dacb68cf04841d44a3e3ed003fe231e44ce5fe234b75cR141.

In any case, I think it'd be better to just send the scalar rather than a numpy array.

jamestiotio replied:
Okay understood. Will add this change.

pcuenca commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -156,6 +196,13 @@ def __call__(
                 plain tuple.
+            callback (`Callable`, *optional*):
+                A function that will be called every `callback_steps` steps during inference. The function will be
+                called with the following arguments: `callback(step: int, timestep: np.ndarray, latents:
```
```suggestion
                called with the following arguments: `callback(step: int, timestep: int, latents:
```

Same as above.

pcuenca commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:
```diff
@@ -53,6 +67,8 @@ def __call__(
         latents: Optional[np.ndarray] = None,
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
+        callback: Optional[Callable[[int, np.ndarray, torch.FloatTensor], None]] = None,
```
```suggestion
        callback: Optional[Callable[[int, int, np.ndarray], None]] = None,
```

`timestep` as mentioned in the Stable Diffusion pipeline.

Regarding the latents, in this case the pipeline works with `numpy` arrays internally, see https://github.com/huggingface/diffusers/pull/654/files.

However, the call to the `scheduler` would return a tensor. I think we should convert to `numpy` for consistency and send that back. Do you agree @anton-l ?

jamestiotio replied:
Okay sure, will add this change as well.

pcuenca commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_onnx.py:
```diff
@@ -140,16 +164,13 @@ def __call__(
-        # scale and decode the image latents with vae
-        latents = 1 / 0.18215 * latents
-        image = self.vae_decoder(latent_sample=latents)[0]
+            # call the callback, if provided
+            if callback is not None and i % callback_steps == 0:
+                callback(i, t, latents)
```
```suggestion
            latents = np.array(latents)
            
            # call the callback, if provided
            if callback is not None and i % callback_steps == 0:
                callback(i, t, latents)
```

As per the explanation above. Do you agree, @anton-l ?

pcuenca reviewed (commented):
I reviewed the main Stable Diffusion pipeline and suggested some changes, essentially to restore the decoding and safety checker to inside the `__call__` function as they were before, since we are just sending the latents to the callback. I also commented a few nits about the timestep sent to the callback, I believe it should just be an `int`.

The same comments would apply to the other pipelines, so I didn't repeat them there. I did add an additional comment in the ONNX pipeline to send the latents as numpy arrays for consistency, but @anton-l is much more knowledgeable about that.

pcuenca commented:
> 2. I think the way the safety checker is handled in this PR is nice.

I'm asking because in my opinion we should send the latents to the callback instead of the decoded image, to make it as general as possible. There are several examples of users leveraging the latents for their work. If that's the case, we won't decode the image and therefore won't run the safety checker for the user.

Downstream applications that expose the intermediate latents should therefore do it themselves. Is that an acceptable solution @patrickvonplaten?

pcuenca commented:
> We just need to add three tests to merge this PR @jamestiotio could you maybe add three test showing intermediate decoding of the pipelines in `test_pipelines.py`?

@jamestiotio would you be able to take care of that? Otherwise I can do it.

pcuenca commented:
By the way, great work @jamestiotio, I liked it a lot that you paid a lot of attention to details and included docstrings and type hints :)

jamestiotio commented:
> If that's the case, we won't decode the image and therefore won't run the safety checker for the user.
> 
> Downstream applications that expose the intermediate latents should therefore do it themselves.

In this case, as mentioned above by @keturn, would it be better to keep the `decode_latents` and `run_safety_checker` functions separated from `__call__` as they are now to allow downstream applications to simply call those functions by themselves? This way, we can minimize code duplication.

> Final question, what do you think about @ahrm's proposal to use a subclassable function (empty in the `StableDiffusionPipeline`) to provide the latents? The advantage is that less parameters would be required in the `__call__` function.

As a user of the library, I am impartial to both implementations (subclassable function and `_call_` function parameters). Since this is more relevant to a stylistic choice and an ease-of-maintenance issue, I will leave this decision to the maintainers (maybe @patrickvonplaten?).

> > We just need to add three tests to merge this PR @jamestiotio could you maybe add three test showing intermediate decoding of the pipelines in `test_pipelines.py`?
> 
> @jamestiotio would you be able to take care of that? Otherwise I can do it.

I should be able to take care of this. Do give me some time to add them.

> By the way, great work @jamestiotio, I liked it a lot that you paid a lot of attention to details and included docstrings and type hints :) From my point of view this is very cool and ready to go when we iron out the final details!

Thank you! 😄

patrickvonplaten commented:
Awesome this looks good to me. Ok to merge whenever @pcuenca is happy with it :-)

jamestiotio commented:
@pcuenca As previously mentioned, I have added the corresponding tests in commit ddbdec7967f3552b9e18c9e2271fdb5014717801.

Do let me know if you require any further changes. Cheers!

pcuenca commented:
Hi @jamestiotio!

Thanks a lot for writing the tests, they look great! However, they fail in my system. I'm seeing two problems:
- The latents are received by the callback in a `cuda` device so they can't be converted to a `numpy` array.
- The computed values do not match the ones you wrote in the expected slice.

Are you testing in GPU too? Otherwise, I can try to fix them myself :)

jamestiotio commented:
Hi @pcuenca, apologies for the previously failing tests. I have fixed the tests accordingly in commit fe05ea2eea78a0789b53e94b86fde6272a49a5a4.

Since my GPU does not seem to possess enough VRAM to be able to load the full model, I have decided to modify the tests slightly:

1. For the `text2img`, `img2img`, and `inpaint` pipelines, I used half-precision weights via the `fp16` branch, used automatic mixed precision via `autocast`, and enabled attention slicing.
2. For the `onnx` pipeline, I changed the test to use `CPUExecutionProvider` instead.

All 4 tests pass on my side. Feel free to modify the values of the `expected_slice` if you decide to use `CUDAExecutionProvider` for the `onnx` pipeline and the full-precision model for the rest.

pcuenca commented:
Hi @jamestiotio don't worry, I'll merge this one and then update the tests so they use standard arguments.

Thanks a lot!

nicollegah commented:
Hi peeps, thanks for your amazing work. Is it possible to use this callback parameter to do masking with the img2img pipeline? And if yes: How would I do that?

patrickvonplaten commented:
Hey @nicollegah,

Could you maybe open a new issue for this one?

nicollegah commented:
Sure! I just did https://github.com/huggingface/diffusers/issues/2073
